core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java

```diff
@@ -273,19 +280,24 @@ void forceSorterToSpill() throws IOException {
       if (maybeSingleFileWriter.isPresent()) {
         // Here, we don't need to perform any metrics updates because the bytes written to this
         // output file would have already been counted as shuffle bytes written.
-        partitionLengths = spills[0].partitionLengths;
-        maybeSingleFileWriter.get().transferMapSpillFile(spills[0].file, partitionLengths);
+        Optional<MapOutputMetadata> maybeMetadata =
+            maybeSingleFileWriter.get().transferMapSpillFile(
+                spills[0].file, spills[0].partitionLengths);
+        mapOutputCommitMessage = maybeMetadata.map(
+            metadata -> MapOutputCommitMessage.of(spills[0].partitionLengths, metadata))
+            .orElse(MapOutputCommitMessage.of(spills[0].partitionLengths));
```

Comment:
I cannot se why `transferMapSpillFile` cannot return a `MapOutputCommitMessage` that would simply this part:
https://github.com/attilapiros/spark/commit/289050e4309b024146867b7b32974ab4746eb570

Reply:
Hm, I think this was originally designed this way because we didn't want the single spill writer to set a list of partition lengths that was different from what was passed into the writer's transfer function. But, maybe we can wrap this with a preconditions check to ensure that the state remains consistent, and that's good enough along with Javadoc.